File: src/server/live-engine.ts
import { EventEmitter } from "node:events";
import { validateEntry } from "../lib/entries";
import { RingBuffer } from "../lib/ring-buffer";
import {
  GAME_SECONDS_PER_TICK,
  LIVE_LINE_NUDGE_CHANCE,
  PREGAME_LINE_NUDGE_CHANCE,
  RESTART_AFTER_TICKS,
  TICK_MS,
  advanceStat,
  clockLabel,
  gameProgress,
  nudgeLine,
  openingStat,
  projectFinal,
  regulationSeconds,
} from "../lib/sim";
import {
  MAX_ENTRIES,
  MAX_TICKS_PER_PROP,
  MAX_WS_CLIENTS,
  type Entry,
  type EntryRequest,
  type MemoryStats,
  type Prop,
  type Snapshot,
  type Tick,
  type WsBatch,
} from "../lib/types";
import { seedGames, seedSlate, type GameSeed } from "./slate";

type GameState = GameSeed;

export class LiveEngine extends EventEmitter {
  private games = new Map<string, GameState>();
  private props = new Map<string, Prop>();
  private history = new Map<string, RingBuffer<Tick>>();
  /** Each live prop's projected final total. Server-side only; never serialized. */
  private projections = new Map<string, number>();
  private entries: Entry[] = [];
  private seq = 0;
  private timer: ReturnType<typeof setInterval> | null = null;
  private entrySeq = 0;
  private finalTicks = 0;

  constructor() {
    super();
    this.setMaxListeners(64);
    this.seedBoard();
  }

  start(): void {
    if (this.timer) return;
    this.timer = setInterval(() => this.tick(), TICK_MS);
    this.timer.unref?.();
  }

  stop(): void {
    if (this.timer) clearInterval(this.timer);
    this.timer = null;
  }

  memory(): MemoryStats {
    let tickCount = 0;
    for (const buf of this.history.values()) tickCount += buf.length;
    return {
      maxTicksPerProp: MAX_TICKS_PER_PROP,
      propCount: this.props.size,
      tickCount,
      clientCap: MAX_WS_CLIENTS,
    };
  }

  snapshot(): Snapshot {
    const props = [...this.props.values()].map((prop) => ({
      ...prop,
      history: this.history.get(prop.id)?.toArray() ?? [],
    }));
    return {
      seq: this.seq,
      generatedAt: Date.now(),
      props,
      memory: this.memory(),
    };
  }

  getProp(id: string) {
    const prop = this.props.get(id);
    if (!prop) return null;
    return { ...prop, history: this.history.get(id)?.toArray() ?? [] };
  }

  health() {
    return {
      ok: true,
      product: "LineLock",
      backend: "node",
      feed: this.timer ? "running" : "stopped",
      seq: this.seq,
      memory: this.memory(),
    };
  }

  listEntries() {
    return this.entries.slice(-20).reverse();
  }

  submitEntry(body: EntryRequest) {
    const result = validateEntry(body, this.props);
    if (!result.ok) return result;
    this.entrySeq += 1;
    const entry: Entry = {
      id: `ent_${this.entrySeq}`,
      createdAt: Date.now(),
      legs: result.legs,
      multiplier: result.multiplier,
      status: "pending",
    };
    this.entries.push(entry);
    if (this.entries.length > MAX_ENTRIES) {
      this.entries.splice(0, this.entries.length - MAX_ENTRIES);
    }
    return { ok: true as const, entry };
  }

  onTick(listener: (batch: WsBatch) => void): void {
    this.on("batch", listener);
  }

  /** Reseeds the slate. Exposed for tests and for the end-of-slate restart. */
  seedBoard(): void {
    this.games.clear();
    this.props.clear();
    this.history.clear();
    this.projections.clear();
    this.finalTicks = 0;

    for (const game of seedGames()) this.games.set(game.id, { ...game });

    for (const seed of seedSlate()) {
      const game = this.games.get(seed.gameId);
      if (!game) continue;
      const prop: Prop = {
        ...seed,
        gameStatus: game.status,
        clock: clockLabel(game.sport, game.status, game.elapsedSec, game.startLabel),
        liveStat: null,
        updatedAt: Date.now(),
      };
      if (game.status !== "scheduled") {
        const projected = projectFinal(prop.line, prop.stat, Math.random);
        this.projections.set(prop.id, projected);
        prop.liveStat = openingStat(
          projected,
          prop.stat,
          gameProgress(game.sport, game.elapsedSec),
          Math.random,
        );
      }
      this.props.set(prop.id, prop);
      const buf = new RingBuffer<Tick>(MAX_TICKS_PER_PROP);
      buf.push({ t: prop.updatedAt, line: prop.line, liveStat: prop.liveStat });
      this.history.set(prop.id, buf);
    }
  }

  /** One simulated tick. Exposed so tests can drive the clock without waiting. */
  tick(): WsBatch | null {
    this.advanceClocks();

    const updates: WsBatch["updates"] = [];
    for (const prop of this.props.values()) {
      const game = this.games.get(prop.gameId);
      if (!game) continue;
      if (!this.advanceProp(prop, game)) continue;
      prop.updatedAt = Date.now();
      const tick: Tick = { t: prop.updatedAt, line: prop.line, liveStat: prop.liveStat };
      this.history.get(prop.id)?.push(tick);
      updates.push({ id: prop.id, prop: { ...prop }, tick });
    }

    if (this.restartIfSlateIsOver()) {
      return this.emitBatch(
        [...this.props.values()].map((prop) => ({
          id: prop.id,
          prop: { ...prop },
          tick: { t: prop.updatedAt, line: prop.line, liveStat: prop.liveStat },
        })),
      );
    }

    if (!updates.length) return null;
    return this.emitBatch(updates);
  }

  private emitBatch(updates: WsBatch["updates"]): WsBatch {
    this.seq += 1;
    const batch: WsBatch = { type: "batch", seq: this.seq, updates };
    this.emit("batch", batch);
    return batch;
  }

  private advanceClocks(): void {
    for (const game of this.games.values()) {
      if (game.status === "scheduled") {
        game.startsInSec -= GAME_SECONDS_PER_TICK;
        if (game.startsInSec <= 0) this.tipOff(game);
      } else if (game.status === "live") {
        const regulation = regulationSeconds(game.sport);
        game.elapsedSec = Math.min(regulation, game.elapsedSec + GAME_SECONDS_PER_TICK);
        if (game.elapsedSec >= regulation) game.status = "final";
      }
    }
  }

  private tipOff(game: GameState): void {
    game.status = "live";
    game.elapsedSec = 0;
    game.startsInSec = 0;
    for (const prop of this.props.values()) {
      if (prop.gameId !== game.id) continue;
      this.projections.set(prop.id, projectFinal(prop.line, prop.stat, Math.random));
      prop.liveStat = 0;
    }
  }

  /** Applies one tick to a prop. Returns whether anything a client can see changed. */
  private advanceProp(prop: Prop, game: GameState): boolean {
    if (game.status === "scheduled") {
      if (Math.random() >= PREGAME_LINE_NUDGE_CHANCE) return false;
      prop.line = nudgeLine(prop.line, prop.openingLine, Math.random);
      return true;
    }

    if (prop.gameStatus === "final") return false;

    const projected = this.projections.get(prop.id) ?? 0;
    prop.liveStat = advanceStat(
      prop.liveStat ?? 0,
      projected,
      prop.stat,
      gameProgress(game.sport, game.elapsedSec),
      Math.random,
    );

    if (game.status === "final") {
      prop.liveStat = projected;
      prop.gameStatus = "final";
      prop.clock = "Final";
      return true;
    }

    prop.gameStatus = "live";
    prop.clock = clockLabel(game.sport, game.status, game.elapsedSec, game.startLabel);
    if (Math.random() < LIVE_LINE_NUDGE_CHANCE) {
      prop.line = nudgeLine(prop.line, prop.openingLine, Math.random);
    }
    return true;
  }

  /** Once every game is final the board would freeze, so reseed after a short hold. */
  private restartIfSlateIsOver(): boolean {
    const over = [...this.games.values()].every((game) => game.status === "final");
    if (!over) {
      this.finalTicks = 0;
      return false;
    }
    this.finalTicks += 1;
    if (this.finalTicks < RESTART_AFTER_TICKS) return false;
    this.seedBoard();
    return true;
  }
}

export const liveEngine = new LiveEngine();

File: backend/app/services/live_engine.rb
# frozen_string_literal: true

require "singleton"

class LiveEngine
  include Singleton

  MAX_TICKS = 48
  MAX_CLIENTS = 32
  MAX_ENTRIES = 100
  TICK_SECONDS = Sim::TICK_SECONDS
  STREAM = "linelock_feed"

  def initialize
    @mutex = Mutex.new
    @games = {}
    @props = {}
    @history = {}
    @projections = {}
    @entries = []
    @seq = 0
    @entry_seq = 0
    @final_ticks = 0
    @timer = nil
    reset_slate!
  end

  def reset_slate!
    @mutex.synchronize do
      @entries = []
      @seq = 0
      @entry_seq = 0
      seed_board
    end
  end

  def start!
    @mutex.synchronize do
      return if @timer

      @timer = Thread.new do
        Thread.current.abort_on_exception = true
        loop do
          sleep TICK_SECONDS
          tick!
        end
      end
    end
  end

  def stop!
    @mutex.synchronize do
      @timer&.kill
      @timer = nil
    end
  end

  def running?
    @mutex.synchronize { !@timer.nil? }
  end

  def health
    {
      "ok" => true,
      "product" => "LineLock",
      "backend" => "rails",
      "feed" => running? ? "running" : "stopped",
      "seq" => @seq,
      "memory" => memory
    }
  end

  def snapshot
    @mutex.synchronize { snapshot_unlocked }
  end

  def get_prop(id)
    @mutex.synchronize do
      prop = @props[id]
      return nil unless prop

      prop.merge("history" => @history[id].to_a)
    end
  end

  def list_entries
    @mutex.synchronize { @entries.last(20).reverse }
  end

  def submit_entry(legs)
    @mutex.synchronize do
      result = EntryValidator.new(legs, @props).validate
      return result unless result[:ok]

      @entry_seq += 1
      entry = {
        "id" => "ent_#{@entry_seq}",
        "createdAt" => Slate.now_ms,
        "legs" => result[:legs],
        "multiplier" => result[:multiplier],
        "status" => "pending"
      }
      @entries << entry
      @entries.shift while @entries.length > MAX_ENTRIES
      { ok: true, entry: entry }
    end
  end

  # One simulated tick. Public so specs can drive the clock without waiting.
  def tick!
    batch = nil
    @mutex.synchronize do
      advance_clocks

      updates = []
      @props.each_value do |prop|
        game = @games[prop["gameId"]]
        next if game.nil?
        next unless advance_prop(prop, game)

        prop["updatedAt"] = Slate.now_ms
        tick = { "t" => prop["updatedAt"], "line" => prop["line"], "liveStat" => prop["liveStat"] }
        @history[prop["id"]].push(tick)
        updates << { "id" => prop["id"], "prop" => prop.dup, "tick" => tick }
      end

      if restart_if_slate_is_over
        updates = @props.values.map do |prop|
          tick = { "t" => prop["updatedAt"], "line" => prop["line"], "liveStat" => prop["liveStat"] }
          { "id" => prop["id"], "prop" => prop.dup, "tick" => tick }
        end
      end

      next if updates.empty?

      @seq += 1
      batch = { "type" => "batch", "seq" => @seq, "updates" => updates }
    end

    return unless batch

    ActionCable.server.broadcast(STREAM, batch)
    SocketHub.broadcast(batch.to_json)
    batch
  end

  private

  def seed_board
    @games = {}
    @props = {}
    @history = {}
    @projections = {}
    @final_ticks = 0

    Slate.games.each { |game| @games[game["id"]] = game }

    Slate.seed.each do |seed|
      game = @games[seed["gameId"]]
      next if game.nil?

      prop = seed.merge(
        "gameStatus" => game["status"],
        "clock" => Sim.clock_label(game["sport"], game["status"], game["elapsedSec"], game["startLabel"]),
        "liveStat" => nil,
        "updatedAt" => Slate.now_ms
      )
      unless game["status"] == "scheduled"
        projected = Sim.project_final(prop["line"], prop["stat"])
        @projections[prop["id"]] = projected
        progress = Sim.game_progress(game["sport"], game["elapsedSec"])
        prop["liveStat"] = Sim.opening_stat(projected, prop["stat"], progress)
      end

      @props[prop["id"]] = prop
      buf = RingBuffer.new(MAX_TICKS)
      buf.push({ "t" => prop["updatedAt"], "line" => prop["line"], "liveStat" => prop["liveStat"] })
      @history[prop["id"]] = buf
    end
  end

  def advance_clocks
    @games.each_value do |game|
      case game["status"]
      when "scheduled"
        game["startsInSec"] -= Sim::GAME_SECONDS_PER_TICK
        tip_off(game) if game["startsInSec"] <= 0
      when "live"
        regulation = Sim.regulation_seconds(game["sport"])
        game["elapsedSec"] = [ game["elapsedSec"] + Sim::GAME_SECONDS_PER_TICK, regulation ].min
        game["status"] = "final" if game["elapsedSec"] >= regulation
      end
    end
  end

  def tip_off(game)
    game["status"] = "live"
    game["elapsedSec"] = 0
    game["startsInSec"] = 0
    @props.each_value do |prop|
      next unless prop["gameId"] == game["id"]

      @projections[prop["id"]] = Sim.project_final(prop["line"], prop["stat"])
      prop["liveStat"] = 0
    end
  end

  # Applies one tick to a prop. Returns whether anything a client can see changed.
  def advance_prop(prop, game)
    if game["status"] == "scheduled"
      return false if rand >= Sim::PREGAME_LINE_NUDGE_CHANCE

      prop["line"] = Sim.nudge_line(prop["line"], prop["openingLine"])
      return true
    end

    return false if prop["gameStatus"] == "final"

    projected = @projections[prop["id"]] || 0
    progress = Sim.game_progress(game["sport"], game["elapsedSec"])
    prop["liveStat"] = Sim.advance_stat(prop["liveStat"] || 0, projected, prop["stat"], progress)

    if game["status"] == "final"
      prop["liveStat"] = projected
      prop["gameStatus"] = "final"
      prop["clock"] = "Final"
      return true
    end

    prop["gameStatus"] = "live"
    prop["clock"] = Sim.clock_label(game["sport"], game["status"], game["elapsedSec"], game["startLabel"])
    prop["line"] = Sim.nudge_line(prop["line"], prop["openingLine"]) if rand < Sim::LIVE_LINE_NUDGE_CHANCE
    true
  end

  # Once every game is final the board would freeze, so reseed after a short hold.
  def restart_if_slate_is_over
    unless @games.each_value.all? { |game| game["status"] == "final" }
      @final_ticks = 0
      return false
    end

    @final_ticks += 1
    return false if @final_ticks < Sim::RESTART_AFTER_TICKS

    seed_board
    true
  end

  def snapshot_unlocked
    props = @props.values.map do |prop|
      prop.merge("history" => @history[prop["id"]].to_a)
    end
    {
      "seq" => @seq,
      "generatedAt" => Slate.now_ms,
      "props" => props,
      "memory" => memory_unlocked
    }
  end

  def memory
    @mutex.synchronize { memory_unlocked }
  end

  def memory_unlocked
    tick_count = @history.values.sum(&:length)
    {
      "maxTicksPerProp" => MAX_TICKS,
      "propCount" => @props.size,
      "tickCount" => tick_count,
      "clientCap" => MAX_CLIENTS
    }
  end
end
